README.md
# Thor

Thor is WebSocket benchmarking/load generator. There are a lot of benchmarking
tools for HTTP servers. You've got ab, siege, wrk and more. But all these tools
only work with plain ol HTTP and have no support for WebSockets - even if they did
they wouldn't be suitable, as they would be testing short running HTTP requests
instead of long running HTTP requests with a lot of messaging traffic. Thor
fixes all of this.

### Dependencies

Thor requires Node.js to be installed on your system. If you don't have Node.js
installed you can download it from http://nodejs.org or build it from the github
source repository: http://github.com/joyent/node.

Once you have Node.js installed, you can use the bundled package manager `npm` to
install this module:

```
npm install -g git://github.com/iorichina/thor.git
```

The `-g` command flag tells `npm` to install the module globally on your system.

### Usage

```
thor [options] <urls>
socketio [options] <url>[@@vip] [<url>[@@vip]...]
```

Thor can hit multiple URL's at once; this is useful if you are testing your
reverse proxies, load balancers or just simply multiple applications. The url
that you supply to `thor` should be written in a WebSocket compatible format
using the `ws` or `wss` protocols:

```
thor --amount 5000 ws://localhost:8080 wss://localhost:8081
```
or use http/https if and only if ur using socket.io(nodejs/java/etc) in server
```
socketio --amount 5000 http://localhost:8080 https://localhost:8081
```

The snippet above will open up `5000` connections against the regular
`ws://localhost:8080` and also `5000` connections against the *secured*
`wss://localhost:8081` server, so a total of `10000` connections will be made.

One thing to keep in mind is you probably need to bump the amount of file
descriptors on your local machine if you start testing WebSockets. Set the
`ulimit -n` on machine as high as possible. If you do not know how to do this,
Google it.

And the other thing u have to know is that, one ip can only create max to 60 thousands connections,
which is limited by TCP/IP protocal. U can special the [@@one_of_the_machine_vip] after the url.

#### thor Options

```
  Usage: thor [options] ws://localhost

  Options:

    -h, --help                      output usage information
    -A, --amount <connections>      the amount of persistent connections to generate
    -C, --concurrent <connections>  how many concurrent-connections per second
    -M, --messages <messages>       number of messages to be send per connection
    -P, --protocol <protocol>       WebSocket protocol version
    -B, --buffer <size>             size of the messages that are send
    -W, --workers <cpus>            workers to be spawned
    -G, --generator <file>          custom message generators
    -M, --masked                    send the messaged with a mask
    -b, --binary                    send binary messages instead of utf-8
    -T, --runtime <seconds>         timeout to close socket(seconds), default to unlimited and u must stop by ctrl+c
    -V, --version                   output the version number
```

Some small notes about the options:

- `--protocol` is the protocol version number. If you want to use the *HyBi drafts
  07-12* use `8` as argument or if you want to use the *HyBi drafts 13-17*
  drafts which are the default version use `13`.
- `--buffer` should be size of the message in bytes.
- `--workers` as Node.js is single threaded this sets the amount of sub
  processes to handle all the heavy lifting.

#### socketio Options

```
  Usage: socketio [options] http[s]://localhost[@@vip]

  Options:

    -h, --help                                  output usage information
    -A, --amount <connections>                  the amount of persistent connections to generate
    -C, --concurrent <connections>[deprecated]  how many concurrent-connections per second
    -M, --messages <messages>                   number of messages to be send per connection
    -P, --protocol <protocol>                   WebSocket protocol version
    -B, --buffer <size>                         size of the messages that are send
    -W, --workers <cpus>                        workers to be spawned
    -M, --masked                                send the messaged with a mask
    -b, --binary                                send binary messages instead of utf-8
    -T, --runtime <seconds>                     timeout to close socket(seconds), default to unlimited and u must stop by ctrl+c
    -V, --version                               output the version number
```


### Custom messages

Some WebSocket servers have their own custom messaging protocol. In order to
work with those servers we introduced a concept called `generators` a generator
is a small JavaScript file that can output `utf8` and `binary` messages. It uses
a really simple generator by default. 

Checkout https://github.com/observing/thor/blob/master/generator.js for an
example of a generator.

```
thor --amount 1000 --generator <file.js> ws://localhost:8080
```

### Example

```
thor --amount 1000 --messages 100 ws://localhost:8080
```

This will hit the WebSocket server that runs on localhost:8080 with 1000
connections and sends 100 messages over each established connection. Once `thor`
is done with smashing your connections it will generate a detailed report:

```
Thor:                                                  version: 1.0.0

God of Thunder, son of Odin and smasher of WebSockets!

Thou shall:
- Spawn 4 workers.
- Create all the concurrent/parallel connections.
- Smash 1000 connections with the mighty Mjölnir.

The answers you seek shall be yours, once I claim what is mine.

Connecting to ws://localhost:8080

  Opened 100 connections
  Opened 200 connections
  Opened 300 connections
  Opened 400 connections
  Opened 500 connections
  Opened 600 connections
  Opened 700 connections
  Opened 800 connections
  Opened 900 connections
  Opened 1000 connections


Online               15000 milliseconds
Time taken           31775 milliseconds
Connected            1000
Disconnected         0
Failed               0
Total transferred    120.46MB
Total received       120.43MB

Durations (ms):

                     min     mean     stddev  median max
Handshaking          217     5036       4094    3902 14451
Latency              0       215         104     205 701

Percentile (ms):

                      50%     66%     75%     80%     90%     95%     98%     98%    100%
Handshaking          3902    6425    8273    9141    11409   12904   13382   13945   14451
Latency              205     246     266     288     371     413     437     443     701
```

### License

MIT


CHANGELOG.md
### 1.1.2

make things better

### 1.0.1

Add socket.io support

### 1.0.0

Initial release of `thor`.


mjolnir.js
'use strict';

var Socket = require('ws')
  , connections = {}
  , concurrent = 0;

//
// Get the session document that is used to generate the data.
//
var session = require(process.argv[2]);

//
// WebSocket connection details.
//
var masked = process.argv[4] === 'true'
  , binary = process.argv[5] === 'true'
  , protocol = +process.argv[3] || 13;

// collect metics datas
var metrics_datas = {collection:true, datas:[]}
  , statInterval = +process.argv[6] || 60
  , process_send = function(data, task) {
      if (statInterval <= 0 || ('open' == data.type && task.nextTask)) {
        process.send(data);
      }else{
        metrics_datas.datas.push(data);
      }
    }
  , process_sendAll = function(end) {
      if (metrics_datas.datas.length <= 0) {
        return;
      };
      // send all data to parent
      process.send(metrics_datas, null, function clearDatas(err){
        // invoked after the message is sent but before the target may have received it
        if (err) {return;}
        // WARNING: maybe we should use synchronize method here
        metrics_datas.datas = [];
        if (end) {
          process.exit();
        };
      });
    }
  , checkConnectionLength = function(){
      if (Object.keys(connections).length <= 0) {
        process_sendAll(true);
      }
    }
  , workerStatInterval = setInterval(function () {
      process_sendAll();
    }, statInterval * 1000);

process.on('message', function message(task) {
  var now = Date.now();

  //
  // Write a new message to the socket. The message should have a size of x
  //
  if ('write' in task) {
    Object.keys(connections).forEach(function write(id) {
      write(connections[id], task, id);
    });
  }

  //
  // Shut down every single socket.
  //
  if (task.shutdown) {
    Object.keys(connections).forEach(function shutdown(id) {
      connections[id] && connections[id].close();
    });
  }

  // End of the line, we are gonna start generating new connections.
  if (!task.url) return;

  var socket = new Socket(task.url, {
    protocolVersion: protocol,
    localAddress: task.localaddr || null
  });
  socket.last = Date.now();
  var pingInterval = null;

  socket.on('open', function open() {
    process_send({ type: 'open', duration: Date.now() - now, id: task.id, concurrent: concurrent, workerid: process.pid }, task);
    // write(socket, task, task.id);

    if (task.pingInterval && task.pingInterval > 0) {
      pingInterval = setInterval(function ping(id, socket) {
        if(socket){
          if(task.serverEngine && -1 != ['socket.io','engine.io','netty-socketio'].indexOf(task.serverEngine)) {
            socket.send('2');
          }else{
            socket.ping();
          }
        }else{
          delete connections[task.id];
          clearInterval(pingInterval);
          checkConnectionLength();
        }
      }, task.pingInterval * 1000, task.id, socket);
    }
    // As the `close` event is fired after the internal `_socket` is cleaned up
    // we need to do some hacky shit in order to tack the bytes send.
  });

  socket.on('message', function message(data) {
    process_send({
      type: 'message', latency: Date.now() - socket.last, concurrent: concurrent, workerid: process.pid,
      id: task.id
    }, task);

    // Only write as long as we are allowed to send messages
    if (task.messages > 0)
    if (--task.messages) {
      write(socket, task, task.id);
    } else {
      socket.close();
      socket.emit('close');
    }
  });
  socket.on('onMessage', function onMessage(data) {
    process_send({
      type: 'message', latency: Date.now() - socket.last, concurrent: concurrent, workerid: process.pid,
      id: task.id
    }, task);

    // Only write as long as we are allowed to send messages
    if (task.messages > 0)
    if (--task.messages) {
      write(socket, task, task.id);
    } else {
      socket.close();
      socket.emit('close');
    }
  });

  socket.on('close', function close() {
    var internal = socket._socket || {};

    process_send({
      type: 'close', id: task.id, concurrent: --concurrent, workerid: process.pid,
      read: internal.bytesRead || 0,
      send: internal.bytesWritten || 0
    }, task);

    delete connections[task.id];
    clearInterval(pingInterval);
    checkConnectionLength();
  });

  socket.on('error', function error(err) {
    process_send({ type: 'error', message: err.message, id: task.id, concurrent: --concurrent, workerid: process.pid }, task);

    socket.close();
    socket.emit('close');
    delete connections[task.id];
    checkConnectionLength();
  });

  // Adding a new socket to our socket collection.
  ++concurrent;
  connections[task.id] = socket;

  // timeout to close socket
  if (task.runtime && task.runtime > 0) {
    setTimeout(function timeoutToCloseSocket(id, socket) {
      socket.close();
      socket.emit('close');
    }, task.runtime * 1000, task.id, socket);
  }
});

process.on('SIGINT', function () {});
process.on('exit', function () {});

/**
 * Helper function from writing messages to the socket.
 *
 * @param {WebSocket} socket WebSocket connection we should write to
 * @param {Object} task The given task
 * @param {String} id
 * @param {Function} fn The callback
 * @param {String} data
 * @api private
 */
function write(socket, task, id, fn, data) {
  // i thank the generator doesn't make any sense, but just let me do some change and leave it alone
  session[binary ? 'binary' : 'utf8'](data || task.size, function message(err, data) {
    var start = socket.last = Date.now();

    socket.send(data, {
      binary: binary,
      mask: masked
    }, function sending(err) {
      if (err) {
        process_send({ type: 'error', message: err.message, concurrent: --concurrent, workerid: process.pid, id: id }, task);

        socket.close();
        socket.emit('close');
        delete connections[id];
      }

      if (fn) fn(err);
    });
  });
}


session.js
'use strict';

/**
 * Generate a UTF-8 messages that we will be send to a connected client.
 *
 * @async
 * @param {Number} size The specified in bytes for the message.
 * @param {Function} fn The callback function for the data.
 * @public
 */
exports.utf8 = function utf(size, fn) {
  var key = 'utf8::'+ size
    , cached = cache[key];

  // We have a cached version of this size, return that instead.
  if (cached) return fn(undefined, cached);

  cached = cache[key] = new Buffer(size).toString('utf-8');
  fn(undefined, cached);
};

/**
 * Generate a binary message that we will be send to a connected client.
 *
 * @async
 * @param {Number} size The specified in bytes for the message.
 * @param {Function} fn The callback function for the data.
 * @public
 */
exports.binary = function binary(size, fn) {
  var key = 'binary::'+ size
    , cached = cache[key];

  // We have a cached version of this size, return that instead.
  if (cached) return fn(undefined, cached);

  cached = cache[key] = new Buffer(size);
  fn(undefined, cached);
};

//
// The following is not needed to create a session file. We don't want to
// re-create & re-allocate memory every time we receive a message so we cache
// them in a variable.
//
var cache = Object.create(null);


thunderbolt.js
'use strict';

var Socket = require('ws')
  , connections = {};

//
// Get the session document that is used to generate the data.
//
var session = require(process.argv[2]);

//
// WebSocket connection details.
//
var masked = process.argv[4] === 'true'
  , binary = process.argv[5] === 'true'
  , protocol = +process.argv[3] || 13;

process.on('message', function message(task) {
  var now = Date.now();

  //
  // Write a new message to the socket. The message should have a size of x
  //
  if ('write' in task) {
    Object.keys(connections).forEach(function write(id) {
      write(connections[id], task, id);
    });
  }

  //
  // Shut down every single socket.
  //
  if (task.shutdown) {
    Object.keys(connections).forEach(function shutdown(id) {
      connections[id].close();
    });
  }

  // End of the line, we are gonna start generating new connections.
  if (!task.url) return;

  var socket = new Socket(task.url, {
    protocolVersion: protocol
  });

  socket.on('open', function open() {
    process.send({ type: 'open', duration: Date.now() - now, id: task.id });
    write(socket, task, task.id);

    // As the `close` event is fired after the internal `_socket` is cleaned up
    // we need to do some hacky shit in order to tack the bytes send.
  });

  socket.on('message', function message(data) {
    process.send({
      type: 'message', latency: Date.now() - socket.last,
      id: task.id
    });

    // Only write as long as we are allowed to send messages
    if (--task.messages) write(socket, task, task.id);
  });

  socket.on('close', function close() {
    process.send({
      type: 'close', id: task.id,
      read: socket._socket.bytesRead,
      send: socket._socket.bytesWritten
    });
  });

  socket.on('error', function error(err) {
    process.send({ type: 'error', message: err.message, id: task.id });

    socket.close();
    delete connections[task.id];
  });

  // Adding a new socket to our socket collection.
  connections[task.id] = socket;
});

/**
 * Helper function from writing messages to the socket.
 *
 * @param {WebSocket} socket WebSocket connection we should write to
 * @param {Object} task The given task
 * @param {String} id
 * @param {Function} fn The callback
 * @api private
 */
function write(socket, task, id, fn) {
  var start = socket.last = Date.now();

  session[binary ? 'binary' : 'utf8'](task.size, function message(err, data) {
    socket.send(data, {
      binary: binary,
      mask: masked
    }, function sending(err) {
      if (err) {
        process.send({ type: 'error', message: err.message });

        socket.close();
        delete connections[id];
      }

      if (fn) fn(err);
    });
  });
}


worker-socketio.js
'use strict';

var Socket = require('socket.io-client')
  , connections = {}
  , concurrent = 0;

//
// Get the session document that is used to generate the data.
//
var session = require(process.argv[2]);

//
// WebSocket connection details.
//
var masked = process.argv[4] === 'true'
  , binary = process.argv[5] === 'true'
  , protocol = +process.argv[3] || 13;

// collect metics datas
var metrics_datas = {collection:true, datas:[]}
  , statInterval = +process.argv[6] || 60
  , process_send = function(data, task) {
      if (statInterval <= 0 || ('open' == data.type && task.nextTask)) {
        process.send(data);
      }else{
        metrics_datas.datas.push(data);
      }
    }
  , process_sendAll = function(end) {
      if (metrics_datas.datas.length <= 0) {
        return;
      };
      // send all data to parent
      process.send(metrics_datas, null, function clearDatas(err){
        // invoked after the message is sent but before the target may have received it
        if (err) {return;}
        // WARNING: maybe we should use synchronize method here
        metrics_datas.datas = [];
        if (end) {
          process.exit();
        };
      });
    }
  , checkConnectionLength = function(){
      if (Object.keys(connections).length <= 0) {
        process_sendAll(true);
      }
    }
  , workerStatInterval = setInterval(function () {
      process_sendAll();
    }, statInterval * 1000);

process.on('message', function message(task) {
  var now = Date.now();

  //
  // Write a new message to the socket. The message should have a size of x
  //
  if ('write' in task) {
    Object.keys(connections).forEach(function write(id) {
      write(connections[id], task, id);
    });
  }

  //
  // Shut down every single socket.
  //
  if (task.shutdown) {
    Object.keys(connections).forEach(function shutdown(id) {
      connections[id] && connections[id].disconnect();
    });
  }

  // End of the line, we are gonna start generating new connections.
  if (!task.url) return;

  var socket = new Socket(task.url, {
    'force new connection': true,
    reconnection: false,
    timeout: 5000,
    transports: [task.transport],
    protocolVersion: protocol,
    localAddress: task.localaddr || null
  });
  socket.last = Date.now();
  var pingInterval = null;

  socket.on('connect', function open() {
    process_send({ type: 'open', duration: Date.now() - now, id: task.id, concurrent: concurrent, workerid: process.pid }, task);
    // write(socket, task, task.id);

    // As the `close` event is fired after the internal `_socket` is cleaned up
    // we need to do some hacky shit in order to tack the bytes send.
  });

  socket.on('message', function message(data) {
    process_send({
      type: 'message', latency: Date.now() - socket.last, concurrent: concurrent, workerid: process.pid,
      id: task.id
    }, task);

    // Only write as long as we are allowed to send messages
    if (task.messages > 0)
    if (--task.messages) {
      write(socket, task, task.id);
    } else {
      socket.disconnect();
      socket.emit('disconnect');
    }
  });
  socket.on('onMessage', function onMessage(data) {
    process_send({
      type: 'message', latency: Date.now() - socket.last, concurrent: concurrent, workerid: process.pid,
      id: task.id
    }, task);

    // Only write as long as we are allowed to send messages
    if (task.messages > 0)
    if (--task.messages) {
      write(socket, task, task.id);
    } else {
      socket.disconnect();
      socket.emit('disconnect');
    }
  });

  socket.on('disconnect', function close() {
    var internal = {};
    try{
      internal = socket.io.engine.transport.ws._socket || {};
    }catch(e){
      // console.log(socket.io.engine.transport.pollXhr);
    }

    process_send({
      type: 'close', id: task.id, concurrent: --concurrent, workerid: process.pid,
      read: internal.bytesRead || 0,
      send: internal.bytesWritten || 0
    }, task);

    delete connections[task.id];
    clearInterval(pingInterval);
    checkConnectionLength();
  });

  socket.on('error', function error(err) {
    process_send({ type: 'error', message: err.description ? err.description.message : err.message, id: task.id, concurrent: --concurrent, workerid: process.pid }, task);

    socket.disconnect();
    socket.emit('disconnect');
    delete connections[task.id];
    checkConnectionLength();
  });

  // catch ECONNREFUSED
  socket.io.on('connect_error', function(err){
    process_send({ type: 'error', message: err.description ? err.description.message : err.message, id: task.id, concurrent: --concurrent, workerid: process.pid }, task);

    socket.disconnect();
    socket.emit('disconnect');
    delete connections[task.id];
    checkConnectionLength();
  });

  // Adding a new socket to our socket collection.
  ++concurrent;
  connections[task.id] = socket;

  // timeout to close socket
  if (task.runtime && task.runtime > 0) {
    setTimeout(function timeoutToCloseSocket(id, socket) {
      socket.disconnect();
      socket.emit('disconnect');
    }, task.runtime * 1000, task.id, socket);
  }
});

process.on('SIGINT', function () {});
process.on('exit', function () {});

/**
 * Helper function from writing messages to the socket.
 *
 * @param {WebSocket} socket WebSocket connection we should write to
 * @param {Object} task The given task
 * @param {String} id
 * @param {Function} fn The callback
 * @param {String} data
 * @api private
 */
function write(socket, task, id, fn, data) {
  // i thank the generator doesn't make any sense, but just let me do some change and leave it alone
  session[binary ? 'binary' : 'utf8'](data || task.size, function message(err, data) {
    var start = socket.last = Date.now();

    socket.send(data, {
      binary: binary,
      mask: masked
    }, function sending(err) {
      if (err) {
        process_send({ type: 'error', message: err.message, concurrent: --concurrent, workerid: process.pid, id: id }, task);

        socket.disconnect();
        socket.emit('disconnect');
        delete connections[id];
      }

      if (fn) fn(err);
    });
  });
}


metrics.js
'use strict';

var Stats = require('fast-stats').Stats
  , colors = require('colors')
  , sugar = require('sugar')
  , table = require('tab');

/**
 * Metrics collection and generation.
 *
 * @constructor
 * @param {Number} requests The total amount of requests scheduled to be send
 */
function Metrics(requests) {
  this.requests = requests;             // The total amount of requests send

  this.connections = 0;                 // Connections established
  this.disconnects = 0;                 // Closed connections
  this.failures = 0;                    // Connections that received an error

  this.errors = Object.create(null);    // Collection of different errors
  this.timing = Object.create(null);    // Different timings

  this.latency = new Stats();           // Latencies of the echo'd messages
  this.handshaking = new Stats();       // Handshake duration

  this.read = 0;                        // Bytes read
  this.send = 0;                        // Bytes send

  // Start tracking
  this.start();
}

/**
 * The metrics has started collecting.
 *
 * @api public
 */
Metrics.prototype.start = function start() {
  this.timing.start = Date.now();
  return this;
};

/**
 * The metrics has stopped collecting.
 *
 * @api public
 */
Metrics.prototype.stop = function stop() {
  if (this.timing.stop) return this;

  this.timing.stop = Date.now();
  this.timing.duration = this.timing.stop - this.timing.start;
  return this;
};

/**
 * All the connections are established
 *
 * @api public
 */
Metrics.prototype.established = function established() {
  if (this.timing.established) return this;

  this.timing.ready = Date.now();
  this.timing.established = this.timing.ready - this.timing.start;
  return this;
};

/**
 * Log an new error.
 *
 * @param {Object} data The error
 * @api public
 */
Metrics.prototype.error = function error(data) {
  this.failures++;

  var collection = this.errors[data.message];
  if (!collection) this.errors[data.message] = 1;
  else this.errors[data.message]++;

  return this;
};

/**
 * Register a message resposne.
 *
 * @param {Object} data The message details.
 * @api public
 */
Metrics.prototype.message = function message(data) {
  this.latency.push(data.latency);

  return this;
};

/**
 * Register a successful handshake + open.
 *
 * @param {Object} data Handshake details.
 * @api public
 */
Metrics.prototype.handshaken = function handshaken(data) {
  this.connections++;
  this.handshaking.push(data.duration);

  return this;
};

/**
 * The connection has closed.
 *
 * @param {Object} data Close information
 * @api public
 */
Metrics.prototype.close = function close(data) {
  this.disconnections++;
  this.read += data.read;
  this.send += data.send;

  return this;
};

/**
 * Generate a summary of the metrics.
 *
 * @returns {Object} The summary
 * @api public
 */
Metrics.prototype.summary = function summary() {
  var results = new table.TableOutputStream({ columns: [
    { label: '', width: 20 },
    { label: '' }
  ]});

  console.log();
  results.writeRow(['Online', this.timing.established + ' milliseconds']);
  results.writeRow(['Time taken', this.timing.duration + ' milliseconds']);
  results.writeRow(['Connected', this.connections]);
  results.writeRow(['Disconnected', this.disconnects]);
  results.writeRow(['Failed', this.failures]);

  results.writeRow(['Total transferred', this.send.bytes(2)]);
  results.writeRow(['Total received', this.read.bytes(2)]);

  // Up next is outputting the series.
  var handshaking = this.handshaking
    , latency = this.latency
    , hrange = handshaking.range()
    , lrange = latency.range();

  //
  // Generate the width of the columns, based on the length of the longest
  // number. If it's less then the max size of a label, we default to that.
  // After that we also pad the strings with 1 char for extra spacing.
  //
  var width = (lrange[1] > hrange[1] ? lrange[1] : hrange[1]).toString().length;
  if (width < 6) width = 6;
  width++;

  console.log();
  console.log('Durations (ms):');
  console.log();

  table.emitTable({
    columns: [
      { label: '', width: 20 },
      { label: 'min', width: width, align: 'left' },
      { label: 'mean', width: width, align: 'left' },
      { label: 'stddev', width: width, align: 'right' },
      { label: 'median', width: width, align: 'right' },
      { label: 'max', width: width, align: 'left' }
    ],
    rows: [
      [
        'Handshaking',
        hrange[0].toFixed(),
        handshaking.amean().toFixed(),
        handshaking.stddev().toFixed(),
        handshaking.median().toFixed(),
        hrange[1].toFixed()
      ],
      [
        'Latency',
        lrange[0].toFixed(),
        latency.amean().toFixed(),
        latency.stddev().toFixed(),
        latency.median().toFixed(),
        lrange[1].toFixed()
      ]
    ]
  });

  console.log();
  console.log('Percentile (ms):');
  console.log();

  table.emitTable({
    columns: [
      { label: '', width: 20 },
      { label: ' 50%', width: width },
      { label: ' 66%', width: width },
      { label: ' 75%', width: width },
      { label: ' 80%', width: width },
      { label: ' 90%', width: width },
      { label: ' 95%', width: width },
      { label: ' 98%', width: width },
      { label: ' 98%', width: width },
      { label: '100%', width: width },
    ],
    rows: [
      [
        'Handshaking',
        handshaking.percentile(50).toFixed(),
        handshaking.percentile(66).toFixed(),
        handshaking.percentile(75).toFixed(),
        handshaking.percentile(80).toFixed(),
        handshaking.percentile(90).toFixed(),
        handshaking.percentile(95).toFixed(),
        handshaking.percentile(98).toFixed(),
        handshaking.percentile(99).toFixed(),
        handshaking.percentile(100).toFixed()
      ],
      [
        'Latency',
        latency.percentile(50).toFixed(),
        latency.percentile(66).toFixed(),
        latency.percentile(75).toFixed(),
        latency.percentile(80).toFixed(),
        latency.percentile(90).toFixed(),
        latency.percentile(95).toFixed(),
        latency.percentile(98).toFixed(),
        latency.percentile(99).toFixed(),
        latency.percentile(100).toFixed()
      ]
    ]
  });

  //
  // Output more error information, there could be multiple causes on why we
  // failed to send a message.
  //
  if (this.failures) {
    console.log();
    console.log('Received errors:');
    console.log();

    Object.keys(this.errors).forEach(function error(err) {
      results.writeRow([this.errors[err] +'x', err]);
    }, this);
  }

  return this;
};

//
// Expose the metrics constructor.
//
module.exports = Metrics;
